### PR TITLE
Refactor api documents

### PR DESCRIPTION
**Pull request by dhlevi**
Update to include the V2 document endpoints. Includes swagger updates and unit testing. Mainly new additions, however to allow for mocking the Minio service a change has been done to the current minio controller, so that part needs to be thoroughly tested.

Like the project V2 endpoints, I expect these will need a number of updates and additions, and the unit tests will be expanded as other related endpoints are created (comments for public, etc.).

Bringing in everyone as a reviewer to make sure we all get a chance to take a look and review if desired!

See ticket [EE-718](https://bcmines.atlassian.net/browse/EE-718)

### CODE REVIEW

**Comment by KitArmstrong** [comment]
FYI, this now has merge conflicts.

File: api/controllers/documentsV2.js [hunk 1]
```diff
@@ -0,0 +1,577 @@
...
+    try
+    {
+        if (args.swagger.params.hasOwnProperty('docId'))
+        {
+            let documentId = args.swagger.params.docId.value;
+            let userName = args.swagger.params.auth_payload.preferred_username ? args.swagger.params.auth_payload.preferred_username : 'public';
+
+            let document = await documentDAO.fetchDocument(documentId);
+            document = await documentDAO.publishDocument(userName, document);
+            document = documentDAO.documentHateoas(updatedDocument, constants.SECURE_ROLES);
```
Comment (KitArmstrong) [inline api/controllers/documentsV2.js]: `updatedDocument' looks undefined.

File: api/controllers/documentsV2.js [hunk 2]
```diff
@@ -0,0 +1,577 @@
...
+
+            if (document)
+            {
+                let documentMeta = await documentDAO.downloadDocumentGetMeta(constants.SECURE_ROLES, userName, document, fileNameOverride);
+
+                res.setHeader('Content-Length', documentMeta.size);
+                res.setHeader('Content-Type', documentMeta.metaData['content-type']);
+                res.setHeader('Content-Disposition', 'attachment;filename="' + documentMeta.fileName + '"');
+
+                return rp(documentMeta.url).pipe(res);
```
Comment (KitArmstrong) [inline api/controllers/documentsV2.js]: What is 'rp' ?

File: api/dao/documentDAO.js [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+};
+
+exports.createDocument = async function(userName, projectId, comment, uploadedFile, ext, documentDetails, isPublic)
+{
+    let guid = intformat(generator.next(), 'dec');
+    let tempFilePath = uploadDir + guid + "." + ext;
+
+    try 
+    {
+        let virusScanSuccessful = false;
```
Comment (KitArmstrong) [inline api/dao/documentDAO.js]: Could just set to `true` and remove the `else` statement.

File: api/dao/documentDAO.js [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+            // Secure create attributes
+            if (isPublic)
+            {
+                // public user can only create documents from comments
+                document.documentSource = "COMMENT";
+            }
+            else
+            {
+                document.legislation          = parseInt(documentDetails.legislation);
+                // doc.labels = JSON.parse(args.swagger.params.labels.value);
```
Comment (KitArmstrong) [inline api/dao/documentDAO.js]: Remove

File: api/dao/documentDAO.js [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+        if (documentDetails.eaoStatus === 'Published') 
+        {
+            originalDocument.read = constants.SECURE_ROLES.concat(constants.PUBLIC_ROLES);
+        } 
+        else if (documentDetails.eaoStatus === 'Rejected') 
+        {
+            originalDocument.read = constants.SECURE_ROLES;
+        }
+
+        // originalDocument.labels = JSON.parse(args.swagger.params.labels.value);
```
Comment (KitArmstrong) [inline api/dao/documentDAO.js]: Remove

**Comment by a1astair** [comment]
Once these conflicts are resolved I can add my review.

File: api/controllers/documentsV2.js (hunk 1 above)
Comment (dhlevi) [reply api/controllers/documentsV2.js]: Yep, copy-paste. Name is updated

File: api/controllers/documentsV2.js (hunk 2 above)
Comment (dhlevi) [reply api/controllers/documentsV2.js]: request-promise-native. I've updated the name and ensured the import is there

File: api/dao/documentDAO.js (hunk 3 above)
Comment (dhlevi) [reply api/dao/documentDAO.js]: Done!

File: api/dao/documentDAO.js (hunk 4 above)
Comment (dhlevi) [reply api/dao/documentDAO.js]: Done!

File: api/dao/documentDAO.js (hunk 5 above)
Comment (dhlevi) [reply api/dao/documentDAO.js]: Done!

**Comment by dhlevi** [comment]
Nope, minio is mocked so you shouldn't need anything running. My docs unit tests pass, it looks like there's a failure in 'generate.test.js': "Cannot find module 'jasmine' from 'generate.test.js'" Which was likely dragged in when I merged in the latest changes to resolve conflicts.

**Comment by KitArmstrong** [comment]
Fixes look good. This might need a rebase... there are 1315 linting errors. I think 99% of those are already fixed.

**Pull request merged**